crypto/bn/bn_prime.c

```diff
@@ -19,14 +19,46 @@
  */
 #include "bn_prime.h"
 
-static int witness(BIGNUM *w, const BIGNUM *a, const BIGNUM *a1,
-                   const BIGNUM *a1_odd, int k, BN_CTX *ctx,
-                   BN_MONT_CTX *mont);
 static int probable_prime(BIGNUM *rnd, int bits, prime_t *mods);
 static int probable_prime_dh_safe(BIGNUM *rnd, int bits,
                                   const BIGNUM *add, const BIGNUM *rem,
                                   BN_CTX *ctx);
 
+#if BN_BITS2 == 64
+#define BN_DEF(l, h) 0x##h##l##ULL
+#else
+#define BN_DEF(l, h) 0x##l, 0x##h
+#endif
+
+/*
+ * See SP800 89 5.3.3 (Step f)
+ * The product of the set of primes ranging from 3 to 751 */
+static const BN_ULONG small_prime_factors[] = {
+    BN_DEF(3ef4e3e1, c4309333), BN_DEF(cd2d655f, 71161eb6),
+    BN_DEF(0bf94862, 95e2238c), BN_DEF(24f7912b, 3eb233d3),
+    BN_DEF(bf26c483, 6b55514b), BN_DEF(5a144871, 0a84d817),
+    BN_DEF(9b82210a, 77d12fee), BN_DEF(97f050b3, db5b93c2),
+    BN_DEF(4d6c026b, 4acad6b9), BN_DEF(54aec893, eb7751f3),
+    BN_DEF(36bc85c4, dba53368), BN_DEF(7f5ec78e, d85a1b28),
+    BN_DEF(6b322244, 2eb072d8), BN_DEF(5e2b3aea, bba51112),
+    BN_DEF(0e2486bf, 36ed1a6c), BN_DEF(ec0c5727, 5f270460),
+    0x000017b1
```

Comment:
I'm not sure how you calculated that, it's not what I get. I'm not sure if BN_dec2bn() would be more clear.

Reply:
This is what I get.

Give the query _**product of primes 3 to 752**_ to Wolfram Alpha gives the result:

> 1090367704589007566035202161494385722019383400119651874476478760664145697976533387343668263403266024900638505861980470638958322454581550203448421495842767449796261170069732921897400657944793163960886418313690808872680411646815934535605444102983629208422567461571568587963766123806881456648026733413053968363611105

Converting this to hex using bc gives:

> 17B15F270460EC0C572736ED1A6C0E2486BFBBA511125E2B3AEA2EB072D86B322244\\
> D85A1B287F5EC78EDBA5336836BC85C4EB7751F354AEC8934ACAD6B94D6C026BDB5B\\
> 93C297F050B377D12FEE9B82210A0A84D8175A1448716B55514BBF26C4833EB233D3\\
> 24F7912B95E2238C0BF9486271161EB6CD2D655FC43093333EF4E3E1

Converting to lower case, prepending four zeros and converting into blocks of eight gives:

> 3ef4e3e1
> c4309333
> cd2d655f
> 71161eb6
> 0bf94862
> 95e2238c
> 24f7912b
> 3eb233d3
> bf26c483
> 6b55514b
> 5a144871
> 0a84d817
> 9b82210a
> 77d12fee
> 97f050b3
> db5b93c2
> 4d6c026b
> 4acad6b9
> 54aec893
> eb7751f3
> 36bc85c4
> dba53368
> 7f5ec78e
> d85a1b28
> 6b322244
> 2eb072d8
> 5e2b3aea
> bba51112
> 0e2486bf
> 36ed1a6c
> ec0c5727
> 5f270460
> 000017b1

This matches the constants listed.

Reply:
I'd prefer to avoid a string to BN conversion that can be done other than at run time.

To me the base ten string `1090367704589007566035202161494385722019383400119651874476478760664145697976533387343668263403266024900638505861980470638958322454581550203448421495842767449796261170069732921897400657944793163960886418313690808872680411646815934535605444102983629208422567461571568587963766123806881456648026733413053968363611105` is about as unintelligible as the hex string or hex constants.

I comment saying how it was calculated might be useful however.  Or even the script or process that produced the number.

Reply:
NIST has a problem in SP 800-89.  The number they give is the product of primes p ∈ [3, 751).  I.e. 751 is not included in their product.  From the text, it seems clear that 751 was intended to be included.

I think it would be faster to test each of the primes in turn (132 divisions by small integers) rather than doing a bignum GCD.

Faster again would be multiplying several of the small primes together but staying within an _int_ doing a modulo reduction and then applying the GCD algorithm to two _ints_.  The first few values for 32 bit _unsigned integers_ would then be:

> 3234846615       = product of 3 5 7 11 13 17 19 23 29
> 95041567        = product of 31 37 41 43 47
> 907383479       = product of 53 59 61 67 71
> 4132280413      = product of 73 79 83 89 97
> 121330189       = product of 101 103 107 109

Reply:
I calculated it initially by writing a small test using part of the bn_primes.h primes[].
(It does this in bn_internal_test to check the constant is correct - which I then pulled out of gdb).

Reply:
I've lodged a query with NIST about their version of the constant.

Reply:
I used the value from the document, so of course I got something else.

Reply:
For completeness: NIST responded saying that they're planning on revisiting this document next fiscal year and the error will be corrected then.

Reply:
I have added a comment indicating that the NIST document is wrong currently.
Also added a comment to say how it was generated.